tests/test_web.py

```diff
@@ -199,3 +218,30 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
     http.post("/invoice", json=email_invoice_request_body)
 
     assert server.starttls.call_count == 0
+
+
+@pytest.mark.parametrize(
```

Comment:
Is there a test for when these env vars aren't set?

Reply:
I thought about it, basically the rest of the tests do this, they would go to default values and as proof the tests still pass.
If I were to test with the defaults in the same tests, it would require another expected_string and a template dir other than tests.

Reply:
Can you write a test that fails?

Reply:
pytest-cov also confirmed that the entire config file is tested including default values. I had a hard time understanding what test to add here. Best I could do is add a new test which mocks jinja2.Environment and checks that it's PackageLoader and get_template params are the defaults of those env variables.

Reply:
`pytest-cov` uses line coverage by default. You can get `coverage` to do a bit better using branch coverage: https://stackoverflow.com/questions/75719802/test-code-and-branch-coverage-simultanously-with-pytest. 

But that's not the entire story either. Branch coverage doesn't cover all cases. There's something called [mutation testing](https://medium.com/analytics-vidhya/unit-testing-in-python-mutation-testing-7a70143180d8) which is halfway decent, but too complex. 

And then there's TDD. Where you have to simply write a test that fails. If you can't write a test that fails, there's nothing left to cover.

Reply:
How about making the template used in tests different from the default? Does not specifying the template related variables still yield the expected result?

Reply:
With the initial test, I specify the test templates which are different from the default ones and an expected_body for the test templates. I can't use the same test as I would need another expected_body for the default templates.
Not specifying the template related variables yields the intended results in all tests as they still pass.

I tried with a test where I specify wrong environment variables values  such as "bad_directory" to check that different errors like TemplateNotFound, ModuleNotFound, etc, are raised but that didn't make much sense to me so I didn't add it.